platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/operations/TestListenerBuildOperationAdapter.java

```diff
@@ -67,15 +67,23 @@ public void started(final TestDescriptorInternal testDescriptor, TestStartEvent
 
     @Override
     public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
-        long currentTime = clock.getCurrentTime();
         InProgressExecuteTestBuildOperation runningOp = runningTests.remove(testDescriptor);
-        listener.finished(runningOp.descriptor, new OperationFinishEvent(runningOp.startTime, currentTime, testResult.getException(), new Result(testResult)));
+        // Adjusted to not use current time, as that may be wrong in the context of remote test execution. The time given by the test result is more reliable.
+        long endTime = completeEvent.getEndTime();
+        if (runningOp.startTime > endTime) {
+            endTime = runningOp.startTime;
+        }
+        listener.finished(runningOp.descriptor, new OperationFinishEvent(runningOp.startTime, endTime, testResult.getException(), new Result(testResult)));
     }
 
     @Override
     public void output(final TestDescriptorInternal testDescriptor, final TestOutputEvent event) {
+        // TODO We may need to not use current time here. We may need a time on the output event.
```

Comment:
It looks like we do the same for started too.  Any idea why TestStartEvent has a time, but we don't use it?

Reply:
I'm not sure. It doesn't look like it was ever explained in https://github.com/gradle/gradle/pull/3298.